actors/datacap/src/lib.rs

```diff
@@ -458,63 +483,63 @@ impl ActorCode for Actor {
         // - the serialize method can't do BigInts
         match FromPrimitive::from_u64(method) {
             Some(Method::Constructor) => {
-                Self::constructor(rt, cbor::deserialize_params(params)?)?;
+                Self::constructor(rt, args)?;
```

Comment:
Hm. This is really annoying. Can we keep the decoding on this side? And maybe decode into some form of "void" type?

Reply:
Ok, solved it:

```rust
use std::marker::PhantomData;

struct Block;
struct Error;

trait Deserialize {
    fn deserialize(value: Block) -> Result<Self, Error>
    where
        Self: Sized;
}

struct SomeArgs;

impl Deserialize for SomeArgs {
    fn deserialize(_: Block) -> Result<Self, Error>
    where
        Self: Sized,
    {
        Ok(SomeArgs)
    }
}

trait Dispatch {
    type Output;
    fn call(self, args: Option<Block>) -> Self::Output;
}

struct Dispatcher<F, A> {
    func: F,
    _marker: PhantomData<fn(A)>,
}

impl<F, A> Dispatcher<F, A> {
    const fn new(f: F) -> Self {
        Dispatcher {
            func: f,
            _marker: PhantomData,
        }
    }
}

fn dispatch<F, A, R>(func: F, arg: Option<Block>) -> Result<R, Error>
where
    Dispatcher<F, A>: Dispatch<Output = Result<R, Error>>,
{
    Dispatcher::new(func).call(arg)
}

impl<F, R> Dispatch for Dispatcher<F, ()>
where
    F: FnOnce() -> Result<R, Error>,
{
    type Output = Result<R, Error>;

    fn call(self, args: Option<Block>) -> Result<R, Error> {
        match args {
            None => (self.func)(),
            Some(_) => Err(Error),
        }
    }
}

impl<F, R, T> Dispatch for Dispatcher<F, (T,)>
where
    F: FnOnce(T) -> Result<R, Error>,
    T: Deserialize,
{
    type Output = Result<R, Error>;

    fn call(self, args: Option<Block>) -> Result<R, Error> {
        match args {
            None => Err(Error),
            Some(arg) => (self.func)(T::deserialize(arg)?),
        }
    }
}

fn with_arg(foo: SomeArgs) -> Result<(), Error> {
    panic!()
}

fn without_arg() -> Result<(), Error> {
    panic!()
}

fn main() {
    let _ = dispatch(with_arg, Some(Block));
    let _ = dispatch(without_arg, None);
}
```